Federation Provider for Azure DNS and related updates to vendored SDKs

xtophs opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
Fixes Issue #44874 

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note

The federation provider for AzureDNS allows for using AzureDNS to manage DNS records for federated kubernetes clusters. The clusters do not have to run in Azure.

Running the provider requires

	• an active Azure subscription with a DNS zone configured. See Azure Docs on how to [configure a DNS Zone](https://docs.microsoft.com/en-us/azure/dns/dns-zones-records).
	• Service Principal credentials with permissions to manage the DNS Zone

To configure federation, you follow the steps at: https://kubernetes.io/docs/tasks/federation/set-up-cluster-federation-kubefed/

Note, that at this time, a cluster created with acs-engine may need kubernetes RBAC (not Azure RBAC) configured first. See detailed instructions link below.

The DNS provider needs a configuration file with the information to access the Azure Subscription. The format of the config file is:

[Global]
subscription-id = [Azure Subscription ID]
tenant-id = [Azure Tenant ID where the Service Principal is created]
client-id = [Service Principal Client ID]
secret = [Service Principal Secret]
resourceGroup = [Azure Resource Group where the DNS Zone is provisioned]

You can get Subscription ID and Tenant ID by running the following Azure CLI 2.0 command:

`az account show`


You can create a Service Principal with write access to the resource group:

`az ad sp create-for-rbac -n "http://MyApp" --role contributor --scopes /subscriptions/[subscription id]/resourceGroups/[resource group name]`

With the provider config file, you can initialize a federation:

`kubefed init myfederation --host-cluster-context=[cluster name] --dns-provider="azure-azuredns" --dns-zone-name="[dns Zone]" --dns-provider-config=[path to provider config file]`

Join a federation:

`kubefed join [your cluster name] --host-cluster-context=[your cluster name] --cluster-context=[your cluster name]`

And verify that the cluster was joined:

kubectl get clusters
NAME STATUS AGE
fed-west Ready 13s

For complete instructions setting up federated clusters in Azure see: https://github.com/xtophs/k8s-setup-federation-cluster
```

xtophs committed:
Federation Provider for Azure DNS and related updates to vendored SDKs

Federation Provider for Azure DNS and related updates to vendored SDKs

xtophs commented:
Thanks.

I followed the steps at https://github.com/kubernetes/community/blob/master/contributors/devel/godep.md to update dependencies azure-sdk-for-go and go-autorest.

Is there something I should to differently?

From: Tim Hockin<mailto:notifications@github.com>
Sent: Tuesday, August 22, 2017 9:07 AM
To: kubernetes/kubernetes<mailto:kubernetes@noreply.github.com>
Cc: Christoph Schittko<mailto:christoph@bracketangles.net>; Mention<mailto:mention@noreply.github.com>
Subject: Re: [kubernetes/kubernetes] Federation Provider for Azure DNS and related updates to vendored SDKs (#51026)


Can someone from @kubernetes/sig-federation-pr-reviews<https://github.com/orgs/kubernetes/teams/sig-federation-pr-reviews> please own this?

@mikedanese<https://github.com/mikedanese> @ixdy<https://github.com/ixdy> - PTAL at why all the BUILD files under vendor/ are regenerated?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/kubernetes/kubernetes/pull/51026#issuecomment-323928197>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AIXXMa3uA3ize95SmGu3dcxKVibX_Ujvks5sanAagaJpZM4O9VtT>.

thockin commented:
If there's a problem with BUILD files, it's not your problem :)  They
should not need to regen for a PR such as this.

On Mon, Aug 21, 2017 at 11:56 PM, k8s-ci-robot <notifications@github.com>
wrote:

> @xtophs <https://github.com/xtophs>: Reiterating the mentions to trigger
> a notification:
> @kubernetes/sig-federation-pr-reviews
> <https://github.com/orgs/kubernetes/teams/sig-federation-pr-reviews>.
>
> In response to this
> <https://github.com/kubernetes/kubernetes/pull/51026#issuecomment-323936464>
> :
>
> Thanks.
>
> I followed the steps at https://github.com/kubernetes/
> community/blob/master/contributors/devel/godep.md to update dependencies
> azure-sdk-for-go and go-autorest.
>
> Is there something I should to differently?
>
> From: Tim Hockinmailto:notifications@github.com <notifications@github.com>
> Sent: Tuesday, August 22, 2017 9:07 AM
> To: kubernetes/kubernetesmailto:kubernetes@noreply.github.com
> <kubernetes@noreply.github.com>
> Cc: Christoph Schittkomailto:christoph@bracketangles.net
> <christoph@bracketangles.net>; Mentionmailto:mention@noreply.github.com
> <mention@noreply.github.com>
> Subject: Re: [kubernetes/kubernetes] Federation Provider for Azure DNS and
> related updates to vendored SDKs (#51026
> <https://github.com/kubernetes/kubernetes/pull/51026>)
>
> Can someone from @kubernetes/sig-federation-pr-reviews
> <https://github.com/orgs/kubernetes/teams/sig-federation-pr-reviews>
> https://github.com/orgs/kubernetes/teams/sig-federation-pr-reviews please
> own this?
>
> @mikedanese <https://github.com/mikedanese>https://github.com/mikedanese
> @ixdy <https://github.com/ixdy>https://github.com/ixdy - PTAL at why all
> the BUILD files under vendor/ are regenerated?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub#51026 (comment)
> <https://github.com/kubernetes/kubernetes/pull/51026#issuecomment-323928197>,
> or mute the threadhttps://github.com/notifications/unsubscribe-auth/
> AIXXMa3uA3ize95SmGu3dcxKVibX_Ujvks5sanAagaJpZM4O9VtT.
>
> Instructions for interacting with me using PR comments are available here
> <https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md>.
> If you have questions or suggestions related to my behavior, please file an
> issue against the kubernetes/test-infra
> <https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:>
> repository.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/51026#issuecomment-323936487>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVOSVJgxEu3_8wH9gr_DqyYHu9VAxks5sanuHgaJpZM4O9VtT>
> .
>

ixdy commented:
It looks like this was originally created a week ago, before #50807 went in, which fixed some issues with regenerating the `BUILD` files under `vendor/`. I also updated https://github.com/kubernetes/community/blob/master/contributors/devel/godep.md with some fixes to the process, which likely also was after you'd generated this. (Namely, after removing `vendor/` and running `hack/godep-save.sh`, you need to touch `vendor/BUILD`.

thockin commented:
So, they should rebase?

On Tue, Aug 22, 2017 at 10:45 AM, Jeff Grafton <notifications@github.com>
wrote:

> It looks like this was originally created a week ago, before #50807
> <https://github.com/kubernetes/kubernetes/pull/50807> went in, which
> fixed some issues with regenerating the BUILD files under vendor/. I also
> updated https://github.com/kubernetes/community/blob/master/
> contributors/devel/godep.md with some fixes to the process, which likely
> also was after you'd generated this. (Namely, after removing vendor/ and
> running hack/godep-save.sh, you need to touch vendor/BUILD.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/51026#issuecomment-324100823>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVOCqzYrn8YejEoAGOjp5jH8otBzjks5saxPCgaJpZM4O9VtT>
> .
>

quinton-hoole commented:
@thockin Yup, I can review, or find someone to.

ixdy commented:
yeah, if this gets rebased (and then run `touch vendor/BUILD` followed by `hack/update-bazel.sh`) this PR should look more reasonable.

xtophs commented:
Let's see what happens. I did:

```
git checkout master
git fetch upstream master
git merge upstream/master
git checkout azuredns-federation
git rebase master
# resolve conflicts
#verify everything builds 
./hack/dev-push-hyperkube.sh
git add vendor/github.com/google/cadvisor/pages/static/BUILD vendor/golang.org/x/text/cases/BUILD vendor/BUILD
git pull
# resolve conflicts again
git commit
git push origin azuredns-federation
```
then read @ixdy suggestion so I did:

```
touch vendor/BUILD
hack/update-bazel.sh
git commit -a
git push origin azuredns-federation
```

Hope this makes things easier

ixdy commented:
It looks like you may have merged upstream/master into your fork, rather than rebasing on upstream/master, so you have a lot of commits showing up in this change.

If you rebase on upstream/master, the huge list of commits should go away.

Also, `touch vendor/BUILD` should no longer be necessary.

xtophs commented:
@ixdy I thought this looked odd. It doesn't let me push without a pull (merge) after `git rebase upstream/master`

Any chance you could ping me and walk me through cleaning this up? TIA

ixdy commented:
After rebasing, you'll need to force push (`git push -f`) to your remote branch on github.

My usual workflow looks something like
```console
$ git checkout master
$ git fetch upstream
$ git merge --ff-only upstream/master
$ git checkout -b some-dev-branch
[work on change]
$ git push
[open PR for review, make more changes]
$ git push  # or push -f, if I squashed or amended any commits
```

If there are merge conflicts, or I want to pull upstream changes into my dev branch:
```console
$ git checkout some-dev-branch
$ git fetch upstream
$ git rebase upstream/master
$ git push -f
```

xtophs commented:
Thanks Much … I didn’t dare to -f on something like k8s without prior ok since Linus apparently isn’t a fan 😉

https://stackoverflow.com/a/432518/8093654

You should see my cleaned up, squashed and rebased attempt shortly.

From: Jeff Grafton<mailto:notifications@github.com>
Sent: Tuesday, August 29, 2017 9:41 PM
To: kubernetes/kubernetes<mailto:kubernetes@noreply.github.com>
Cc: Christoph Schittko<mailto:christoph@bracketangles.net>; Mention<mailto:mention@noreply.github.com>
Subject: Re: [kubernetes/kubernetes] Federation Provider for Azure DNS and related updates to vendored SDKs (#51026)


After rebasing, you'll need to force push (git push -f) to your remote branch on github.

My usual workflow looks something like

$ git checkout master

$ git fetch upstream

$ git merge --ff-only upstream/master

$ git checkout -b some-dev-branch

[work on change]

$ git push

[open PR for review, make more changes]

$ git push  # or push -f, if I squashed or amended any commits

If there are merge conflicts, or I want to pull upstream changes into my dev branch:

$ git checkout some-dev-branch

$ git fetch upstream

$ git rebase upstream/master

$ git push -f

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/kubernetes/kubernetes/pull/51026#issuecomment-325775429>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AIXXMfH1nauj3mGFiT-Z9eXnbx0qaB0oks5sdGhBgaJpZM4O9VtT>.

ixdy commented:
For pushing to a git repo that others have cloned (like the upstream kubernetes/kubernetes repo), you would not want to force-push, since that'll break history for everyone, but for your own private fork it's pretty much the only way.

GitHub kinda breaks the git model in a number of ways.

xtophs commented:
This looks much better. Thanks @ixdy

ixdy commented:
ok, I am unsubscribing from notifications now and letting others review this PR. if you run into more build issues, you can @ me again.

@xtophs it also looks like you need to sign the CLA.

xtophs commented:
CLA is signed now. How do we get the CLA label removed?

xtophs commented:
I rebased and reverified. Let me know if there's anything else I need to do to move it forward.

quinton-hoole commented:
@xtophs Thanks for your patience.  I'll start reviewing this PR now.

quinton-hoole commented on federation/pkg/dnsprovider/providers/azure/azuredns/stubs/azurednsapi.go:
```diff
@@ -0,0 +1,189 @@
+	if ok {
+		delete(a.recordSets, key)
+	} else {
+		// Deleting non-existant item. Some of the tests do that
```
In that case better to call this method "EnsureRecordsetDeleted".
Given that this method never returns anything other than an explicit nil for error, why have it as a return value?

xtophs replied:
This was only a mock. It's not supposed to do much.

quinton-hoole commented on federation/pkg/dnsprovider/providers/azure/azuredns/zones.go:
```diff
@@ -0,0 +1,102 @@
+func (zones Zones) List() ([]dnsprovider.Zone, error) {
+	svc := zones.impl.service
+
+	azZoneList, err := svc.ListZones()
```
Does ListZones return an empty list or an error if there are no zones.  The rest of the stack expects an empty list.  Worth checking.

xtophs replied:
Verified. API does create an empty list

quinton-hoole commented on federation/pkg/dnsprovider/providers/azure/azuredns/zones.go:
```diff
@@ -0,0 +1,102 @@
+	errval := <-err
+	if errval != nil {
+		// TODO: Fix go azure sdk version
+		// Is now async
```
Yes, this might cause some race conditions.  I think the other providers are synchronous (for better or for worse).

xtophs replied:
ensured sync execution. Removed //TODO comment

quinton-hoole commented on staging/src/k8s.io/client-go/Godeps/Godeps.json:
```diff
@@ -16,19 +16,19 @@
 		{
 			"ImportPath": "github.com/Azure/go-autorest/autorest",
-			"Rev": "58f6f26e200fa5dfb40c9cd1c83f3e2c860d779d"
+			"Rev": "48a8154950f853ae1f2a9c409e0eb1fb71119bc8"
```
It would probably be best to split the vendoring of a new version of the dependency into a separate PR.

quinton-hoole replied:
Same for each new/updated dependency.  A separate PR for each one please.  It makes them more visible and easier to review and roll back if necessary,.

xtophs replied:
I tried to piggy back off a recent commit that updated go-autorest and only add github.com/azure/azure-sdk-for-go/arm/dns, but apparently following the steps at https://github.com/kubernetes/community/blob/master/contributors/devel/godep.md seems to update go-autorest and azure-sdk-for-go. 

Would you still like me to break those out?

irfanurrehman commented:
/ok-to-test

irfanurrehman commented:
***FYI***
sig-multicluster is planning a cutover of federation code from core kubernetes to its own repo (k8s.io/kubernetes/federation), and the deadline to merge PRs was posted here. In case this PR misses the earlier communicated window, you would need to re-raise the PR against k8s.io/kubernetes/federation; once the cutover completes (ideally next week). 
@xtophs, this seems to be waiting for the review fixes and some signals that relevant testing is done for AZURE. Will you be able to handle this by the beginning of next week?
cc @quinton-hoole 
cc @kubernetes/sig-multicluster-misc

xtophs commented:
Running e2e tests before next weekend is easy. I can probably also issue separate PRs for the vendored SDKs over the weekend, but would they be merged before the cutover?


From: irfanurrehman<mailto:notifications@github.com>
Sent: Thursday, October 19, 2017 9:44 PM
To: kubernetes/kubernetes<mailto:kubernetes@noreply.github.com>
Cc: Christoph Schittko<mailto:christoph@bracketangles.net>; Mention<mailto:mention@noreply.github.com>
Subject: Re: [kubernetes/kubernetes] Federation Provider for Azure DNS and related updates to vendored SDKs (#51026)


FYI
sig-multicluster is planning a cutover of federation code from core kubernetes to its own repo (k8s.io/kubernetes/federation), and the deadline to merge PRs was posted here. In case this PR misses the earlier communicated window, you would need to re-raise the PR against k8s.io/kubernetes/federation; once the cutover completes (ideally next week).
@xtophs<https://github.com/xtophs>, this seems to be waiting for the review fixes and some signals that relevant testing is done for AZURE. Will you be able to handle this by the beginning of next week?
cc @quinton-hoole<https://github.com/quinton-hoole>
cc @kubernetes/sig-multicluster-misc<https://github.com/orgs/kubernetes/teams/sig-multicluster-misc>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/kubernetes/kubernetes/pull/51026#issuecomment-338015699>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AIXXMWuQ7J3OJrB4bBGrARFBqGTgPXE5ks5st6amgaJpZM4O9VtT>.

irfanurrehman commented:
> Running e2e tests before next weekend is easy. I can probably also issue separate PRs for the vendored SDKs over the weekend, but would they be merged before the cutover?

This seems to be a large piece of code and has already gone through some scrutiny. I would suggest having this merged here, before the cutover (And then carried over with rest of the code during cutover). But as I mentioned, that can happen, if it reaches the desired good shape as late as early next week.

irfanurrehman commented:
@xtophs please update if you would be able to complete this within a day or two (As listed earlier we are sort of already in the freeze period for federation related code)

xtophs commented:
Working on this now. Investigating errors now. Make verify worked when I last committed. More shortly.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

From: irfanurrehman<mailto:notifications@github.com>
Sent: Monday, October 23, 2017 4:25 AM
To: kubernetes/kubernetes<mailto:kubernetes@noreply.github.com>
Cc: Christoph Schittko<mailto:christoph@bracketangles.net>; Mention<mailto:mention@noreply.github.com>
Subject: Re: [kubernetes/kubernetes] Federation Provider for Azure DNS and related updates to vendored SDKs (#51026)


@xtophs<https://github.com/xtophs> please update if you would be able to complete this within a day or two (As listed earlier we are sort of already in the freeze period for federation related code)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/kubernetes/kubernetes/pull/51026#issuecomment-338598605>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AIXXMZ0apokTcZtmwFq7M9rFDZoBD4eJks5svFoggaJpZM4O9VtT>.

marun commented:
@xtophs Please split out of the vendoring changes into a separate commit.  Given one of the updated dependencies is to docker, I'm assuming ``sig/node`` should look at the proposed dep update. 

/sig node

xtophs commented:
There should be no changes to docker. The only changes are to github.com/Azure/* dependencies.

There were some recent changes to github.com/Azure/go-autorest that are probably all I need. I’m currently rebasing and only then only add a *new* dependency to github.com/Azure/azure-sdk-for-go/arm/dns. There should be no other dependency change.

Does that still require a separate PR?

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

From: Maru Newby<mailto:notifications@github.com>
Sent: Monday, October 23, 2017 12:23 PM
To: kubernetes/kubernetes<mailto:kubernetes@noreply.github.com>
Cc: Christoph Schittko<mailto:christoph@bracketangles.net>; Mention<mailto:mention@noreply.github.com>
Subject: Re: [kubernetes/kubernetes] Federation Provider for Azure DNS and related updates to vendored SDKs (#51026)


@xtophs<https://github.com/xtophs> Please split out of the vendoring changes into a separate commit. Given one of the updated dependencies is to docker, I'm assuming sig/node should look at the proposed dep update.

/sig node

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/kubernetes/kubernetes/pull/51026#issuecomment-338733033>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AIXXMc8R23nzduILhp-rEJgFRSULz19bks5svMuSgaJpZM4O9VtT>.

xtophs commented:
addressed all of Quinton's comments.

make verify passes, with the exception of golint. which I believe is based on naming in https://github.com/kubernetes/kubernetes/blob/master/federation/pkg/dnsprovider/dns.go#L95 

It appears that running e2e tests for federation on Azure isn't automated. From what I can tell so far, https://github.com/kubernetes/kubernetes/blob/master/federation/deploy/deploy.sh only works against gcloud.  http://blog.kubernetes.io/2016/07/bringing-end-to-end-kubernetes-testing-to-azure-2.html appears to be outdated.

Is there an easy way to test on Azure or do I need to manually bring up a cluster using the federation images to run federation? 

I've also raised an issue for a bad link in the e2e test doc. I believe build/push-federation-images.sh is not federation/develop/push-federation-images.sh

xtophs commented on federation/pkg/dnsprovider/providers/azure/azuredns/azuredns.go:
```diff
@@ -0,0 +1,82 @@
+}
+
+// newazuredns creates a new instance of an AWS azuredns DNS Interface.
+func newazuredns(config io.Reader) (*Interface, error) {
```
Done

xtophs commented on federation/pkg/dnsprovider/providers/azure/azuredns/helpers.go:
```diff
@@ -0,0 +1,48 @@
+)
+
+const (
+	credentialsPath = "/.azure/credentials.json"
```
Removed

xtophs commented on federation/pkg/dnsprovider/providers/azure/azuredns/interface.go:
```diff
@@ -0,0 +1,168 @@
+func (c Interface) Zones() (dnsprovider.Zones, bool) {
+	return Zones{&c}, true
+}
+
```
Done

xtophs commented on federation/pkg/dnsprovider/providers/azure/azuredns/interface.go:
```diff
@@ -0,0 +1,168 @@
+
+	rrsets := make([]dns.RecordSet, 0)
+
+	result, err := c.rc.ListByDNSZone(c.conf.Global.ResourceGroup,
```
Added error handling

xtophs commented on federation/pkg/dnsprovider/providers/azure/azuredns/interface.go:
```diff
@@ -0,0 +1,168 @@
+// ListZones lists the zones in the configured resource group
+func (c *DNSAPI) ListZones() (dns.ZoneListResult, error) {
+	glog.V(5).Infof("azuredns: Requesting DNS zones")
+	// request all 100 zones. 100 is the current limit per subscription
```
Unfortunately, there are no constants or ways to interrogate the API. Added constants.

xtophs commented on federation/pkg/dnsprovider/providers/azure/azuredns/rrchangeset.go:
```diff
@@ -0,0 +1,140 @@
+
+		if glog.V(5) {
+			switch strings.TrimPrefix(*recType, "Microsoft.Network/dnszones/") {
+			case "A":
```
Done

xtophs commented on federation/pkg/dnsprovider/providers/azure/azuredns/rrset.go:
```diff
@@ -0,0 +1,168 @@
+
+// Name returns the absolute ResourceRecordSet name, i.e. the name includes the zone
+func (rrset ResourceRecordSet) Name() string {
+	if *rrset.impl.Name == "@" {
```
Done

xtophs commented on federation/pkg/dnsprovider/providers/azure/azuredns/rrset.go:
```diff
@@ -0,0 +1,168 @@
+	var rrDatas []string
+
+	switch rrset.Type() {
+	case "A":
```
Done.

xtophs commented on federation/pkg/dnsprovider/providers/azure/azuredns/stubs/azurednsapi.go:
```diff
@@ -0,0 +1,189 @@
+		zones:      make(map[string]*dns.Zone),
+		recordSets: make(map[string][]dns.RecordSet),
+	}
+	api.zones["test.com"] = &dns.Zone{ID: to.StringPtr("zoneID"), Name: to.StringPtr("test.com"), Type: to.StringPtr("ZoneTypes")}
```
switched to .local

xtophs committed:
added Azure SDK for DNS

xtophs commented:
@marun, @irfanurrehman, @quinton-hoole 

Fixed dependencies. No updates of existing dependencies in this PR

marun commented:
/test pull-kubernetes-federation-e2e-gce

marun commented:
@xtophs

 - the vendoring change needs to be in a separate commit, not mixed in with non-vendoring changes
 - please selectively revert the non-functional commit hash changes in Godeps.json

quinton-hoole commented:
@xtophs Yes, you will need to bring up multiple Kubernetes clusters on Azure any way you see fit, deploy the federation control plane into one of them (using kubefed) and register the clusters with the federation (also using kubefed).  Then you can run the federation e2e tests against that federation.  Let me know when you have the tests passing, and I will complete the code review.

marun commented:
@xtophs As per the [godep instructions](https://github.com/kubernetes/community/blob/master/contributors/devel/godep.md#saving-deps), please separate the godep changes into a new commit.  Regardless of whether godep requires references in the code, separation should be possible and would simplify reviewing.

irfanurrehman commented:
> Is there an easy way to test on Azure or do I need to manually bring up a cluster using the federation images to run federation? If not, i would
> 
> build a hyperkube with dev-push-hyperkube.sh,
> follow my setup instructions to build 2 federated clusters and then
> point e2e at the federated clusters to test.
> Is that the right way to go?

Seems like the only way for you as of now.
Some steps with arguments for your reference (you probably already know them).

```
1. Bringup you k8s clusters using your own mechanism.

2. Build and Push hyperkube to private/public registry (the hyperkube binary wraps the federation binaries also)
# REGISTRY=172.17.0.1:5000 VERSION=latest ./hack/dev-push-hyperkube.sh

3. Ensure that your DNS is working fine and get the end point.

3.1 Create your DNS provider conf
# cat <<EOF
# place key value pairs here, if needed by the dns provider code 
# for example dns end points.
EOF > /root/federation-dns-provider.conf

4. Bringup K8s Federation control plane over an existing base (host) cluster
# kubefed init federation --host-cluster-context=<host-cluster-ctx-name-in-kubeconfig> --etcd-persistent-storage=false --dns-provider=azuredns --dns-provider-config=/root/azure-dns-provider.conf --dns-zone-name=<actual-zone-name> --image=172.17.0.1:5000/hyperkube-amd64

{u can additionally use --api-server-service-type=NodePort to not provision LB}


5. Register the k8s cluster
# kubefed join <cluster-context-to-federate> --host-cluster-context=host-cluster-ctx-name-in-kubeconfig --context=federation

6. Verify cluster registeration succeeded
# kubectl --context=federation get clusters

7. Compile e2e binary and Test
# e2e.test --provider=<cloud-provider-name-running-ur-clusters> --kubeconfig /root/.kube/config --federated-kube-context federation --host=https://172.17.0.1:6443 -ginkgo.v=true -ginkgo.focus="\[Feature:Federation\]" -test.v -v 5

--host here is needs ur host cluster API server end point.

Some additional pointers:
federation e2e looks into the kubeconfig to find the federated clusters (this is how the e2e setup is wired in CI jobs) and uses 'federation' prefix to determine if this cluster is federated. So please use a name federation-<any-name> for cluster(s) that u will federate (or join to federation).
```

xtophs commented:
@quinton-hoole 
Would appreciate a couple of pointers. Rebase is currently blocked by failure in `hack/run-in-gopath.sh hack/godep-restore.sh`. Looks like something in the staging directory. Also don't understand the CI/CD errors. All of them seem to occur in places I didn't touch.

```
$ rm -rf _output
$ hack/run-in-gopath.sh hack/godep-restore.sh
+++ [1029 20:05:32] Restoring kubernetes godeps
+++ [1029 20:05:32] Downloading dependencies - this might take a while
+++ [1029 20:07:34] Done

... updating dependencies

$ hack/run-in-gopath.sh hack/godep-restore.sh
+++ [1029 20:07:53] Ensuring prereqs
+++ [1029 20:07:53] Running godep save - this might take a while
godep: /mnt/code/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/staging/src/k8s.io/client-go/plugin/pkg/client/auth/plugins.go:20:1: expected 'STRING', found '<<' (and 2 more errors)
godep: Unable to find dependent package k8s.io/client-go/plugin/pkg/client/auth in context of /mnt/code/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/unversioned
!!! Error in hack/godep-save.sh:69
  Error in hack/godep-save.sh:69. 'GOPATH="${GOPATH}:$(pwd)/staging" godep save "${REQUIRED_BINS[@]}"' exited with status 1
Call stack:
  1: hack/godep-save.sh:69 main(...)
Exiting with status 1
!!! [1029 20:07:54] _tmp/godep-save.9913/vendor exists, restoring it
!!! [1029 20:07:54] _tmp/godep-save.9913/Godeps exists, restoring it
!!! Error in hack/run-in-gopath.sh:33
  Error in hack/run-in-gopath.sh:33. '"${@}"' exited with status 1
Call stack:
  1: hack/run-in-gopath.sh:33 main(...)
Exiting with status 1
```

@marun This is the article I pointed you to. It states that it only saves references found in the build targets, i.e. I can't issue a PR that only adds a new dependency that follows to process in the docs: https://github.com/kubernetes/community/blob/master/contributors/devel/godep.md. Please advise on process to follow. 

Thanks.

irfanurrehman commented:
@xtophs 2 things.
1. The relevant code which ur PR was supposed to be in, has already been removed (https://github.com/kubernetes/kubernetes/pull/53816 already merged)


> @marun This is the article I pointed you to. It states that it only saves references found in the build targets, i.e. I can't issue a PR that only adds a new dependency that follows to process in the docs: https://github.com/kubernetes/community/blob/master/contributors/devel/godep.md. Please advise on process to follow.

2. What @marun probably pointed out earlier was to separate the code that you wrote and the code that was vendored into 2 different commits (but part of the same PR, not different PRs), so that we can cherry-pick your code alone into the move.

Also, there is no point solving problems/resolving rebasing here now. You will need to point this PR to github.com/kubernetes/federation, which might not be much of work as you already have the DNS code ready. You can sync with me (or @marun) on slack to work out issues, if you run into them, especially given that the way we are vendoring the dependencies (including kubernetes/kubernetes) is slightly tricky.

xtophs commented:
@irfanurrehman 

2 things back:

1. Quick heads up would have saved me a couple of hours of work. Please keep in mind that contributors value their time highly and try to keep them informed to avoid wasting time they volunteer to spend on the project.

2. I understand what @marun is saying. I am saying that *it is not possible* to do what he is asking. You cannot to follow the docs and create a PR that only has the new dependency. The doc says so and I verified that it does not work. I am asking to clarify the doc or retract the ask to split into 2 PRs.

irfanurrehman commented:
Came across [this request](https://github.com/kubernetes/federation/issues/130) while cleaning up issues in federation. @xtophs, would you be willing to revive (or help revive) this PR and point it to kubernetes/federation.
